Redo xtrigger docs, with set xtriggers (#820)
cylc/cylc-doc: src/user-guide/writing-workflows/external-triggers.rst

.. _Section External Triggers:

External Triggers (Xtriggers)
=============================

Overview
--------

Xtriggers allow tasks to trigger off of arbitrary external conditions. The
:term:`scheduler` periodically calls a Python function, to check on the
condition, until it returns success to satisfy dependent tasks.

.. note::
   Compared to the older :ref:`Old-Style External Triggers`, xtriggers are
   visible in the graph, they are shared efficiently by all dependent tasks,

>    and they are entirely contained within the workflow.

Author: I think I avoided saying that the new xtriggers are entirely superior, because the old ones are in principle better if your use case can easily support a push mechanism. The new ones are technically superior in every other way though (written in the graph, no need to modify external systems).

>    and they are entirely contained within the workflow.

Author: > I think you mean they are written in the [graph] as opposed to visible in the graph view.

Yes, I suppose I should say "visible in the graph configuration", for now.

Built-in Clock Triggers
-----------------------

Clock xtriggers succeed when the real ("wall clock") time reaches some offset
from the task's cycle point value.

.. note::

   These should be used instead of the older task clock triggers documented in
   :ref:`ClockTriggerTasks`.

The clock xtrigger function signature looks like this:

.. autofunction:: cylc.flow.xtriggers.wall_clock.wall_clock

> The ``offset`` argument is a datetime duration (e.g., ``PT1H`` is 1 hour)

Reviewer: Does this add anything to content of the function doc? I'm not sure it's needed?

Author: I'll remove it.

       [[foo]]
           script = run-foo.sh

> Notice that the short label ``clock_1`` represents the trigger function in the
> graph.

Reviewer: Possible alt?
```suggestion
In the ``[xtriggers]`` section the trigger function is assigned the label ``clock_1``.
``@`` plus this label is what is then used in the graph.
```

> Argument keywords can be omitted if called in the right order, so the

Reviewer: Is this a necessary thing to say?

Author: I think it is necessary, to explain why you can call clock triggers the commonly-used shorthand way, but that for other xtriggers with more arguments, the order has to be right. It should be mentioned in the general section though.

       [[foo]]
           script = run-foo.sh

> However, xtrigger names must adhere to the following rules:

Reviewer: "However" made me re-read the last example (starting at "A zero-offset..") but in reality, the two are unrelated. Consider removing the "However".

Perhaps consider putting this in a `..note::` directive?

Author: Huh, yeah, I suspect that line got repositioned at some point. It should also be in the general section, not clock triggers.

Built-in Workflow State Triggers
--------------------------------

> Workflow-state xtriggers succeed when a target task in another workflow achieves
> a given state or output.

Reviewer: I think it's worth separating the general introduction from the explanation of why the terminology seems backwards (the target is the upstream dependency).

```suggestion
Workflow-state xtriggers succeed when a task in another workflow achieves
a given state or output.

The task in the other workflow is called the target, because it is the target of
the xtrigger.
```

Author: >  from the explanation of why the terminology seems backwards 

I'm not sure I understand this comment. Did you omit some punctuation?

I prefer the original as I'm using "target" as merely an adjective. It's just saying we are targeting a particular task in another workflow - that's the task whose state we are interested. To me that's just normal descriptive English.  Your version sounds more like formalising "target" as special terminology, which wasn't my intention.

> .. literalinclude:: ../../workflows/xtrigger/workflow_state/downstream/flow.cylc

Reviewer: Any particular reason why these `flow.cylc` files are executable?

Author: Nope, news to me. I'll fix it.

   :language: cylc

Try starting the downstream workflow first, then the upstream, and watch what
happens. In each cycle point the ``@upstream`` trigger in the downstream workflow
waits for the upstream task ``foo`` (with the same cycle point) workflow to generate
the "data ready" message.

Some important points to note:

> - The optional interval ``PT10S`` (10 seconds) determines how often the scheduler

Reviewer: Is it worth having a warning here (or below this list) urgining users to consider what a sensible polling interval would be?

Author: I've moved this up to the general section too.

           echo_1 = echo(succeeded=True, hello, 99, point=%(point)s, foo=10)
       [[graph]]
           PT1H = "@echo_1 => foo"
   [runtime]
       [[foo]]
           script = "printenv | grep echo_1"

Run this with ``cylc play --no-detach <workflow>`` and watch your terminal to see
the xtrigger calls. View the task job log with ``cylc cat-log <workflow_id>//1/foo``
to confirm that the dependent task received the xtrigger results.

xrandom
^^^^^^^

The toy ``xrandom`` function sleeps for a configurable amount of time (useful for
testing the effect of a long-running trigger function - which should be avoided)
and has a configurable random chance of success. The function signature is:

.. automodule:: cylc.flow.xtriggers.xrandom
   :members: xrandom, validate
   :member-order: bysource

An example xrandom trigger workflow:

.. literalinclude:: ../../workflows/xtrigger/xrandom/flow.cylc
   :language: cylc

.. _Custom Trigger Functions:

Custom Xtrigger Functions
-------------------------

Xtrigger functions are Python functions with some special requirements.

Requirements
^^^^^^^^^^^^

Xtrigger functions must be compatible with the Python version that runs the

> scheduler (see :ref:`Requirements` for the latest version specification).

Reviewer: Worth adding a note about portability? 

Or adding a note to the document on [Portable workflows](https://github.com/cylc/cylc-doc/blob/master/src/workflow-design-guide/portable-workflows.rst) pointing out that just because you've deployed on Python 3.12 doesn't mean that Walruses and `f'{myvar=}` will work if someone else has Cylc deployed with Python 3.7?

They must return a Tuple of ``(Boolean, Dictionary)``:

  - ``(False, {})`` - failed: trigger condition not met
  - ``(True, results)`` - succeeded: trigger condition met

where ``results`` is a flat (non-nested) dictionary of information to be passed
to dependent tasks - see :ref:`Xtrigger Results`. Each dictionary key must be
valid as an
`environment variable <https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html>`_
name.

> They can take arbitrary positional and keyword arguments, except that the
> ``sequential`` keyword is reserved for :ref:`sequential xtriggers <Sequential Xtriggers>`.

Reviewer: ```suggestion
They can take arbitrary positional and keyword arguments, except for the
keyword ``sequential``, which is reserved for :ref:`sequential xtriggers <Sequential Xtriggers>`.
```

Author: Yep, I prefer your version.

> They cannot store data Pythonically between invocations, because each call

Reviewer: [Don't really care that much, arguable style point] IMO I think "Custom xtrigger functions" or "custom xtriggers", or even just "xtriggers" would be more desirable than "they".

Author: Yes, I was unsure of this myself. I decided a bulleted list of long sentences doesn't look great in this context (it's a more of a book than a powerpoint presentation), but repeating the object on every line seemed too much.

Author: OK I've gone with repetition, as the lesser of two evils.

is executed via a wrapper in a new subprocess. If necessary the filesystem

> could be used for persistent data.

Reviewer: Is this _really_ a pattern we'd recommend?

If they depend on files (say) that might not exist when the function is first
called, just return trigger condition not met (i.e., failed). 

Installation
^^^^^^^^^^^^

We recommend using the ``cylc.xtriggers`` entry point to install the xtrigger
as a Python package - see :ref:`developing.xtrigger.plugins`.

Otherwise, e.g., for installing custom xtriggers under your own user account,
xtrigger functions must be:

  - defined in a module with the same name as the function
  - located in:
    - ``<workflow-dir>/lib/python/``;
    - or anywhere in your ``$CYLC_PYTHONPATH``

Custom xtrigger module can also provide a ``validate`` function for checking
configured arguments, see
:ref:`user-guide.xtrigger-validation-functions` for details.

Passing Workflow Parameters to Xtrigger Functions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Workflow and task parameters can be passed to function arguments from the
workflow configuration via the following string templates. Task parameters
affect :ref:`xtrigger specificity <Xtrigger specificity>`.

.. spelling:word-list::

>    vv

Reviewer: ?

of an xtrigger validation function.

Xtrigger Limitations
--------------------

The following issues may be addressed in future Cylc releases:

> - Trigger names cannot be used in conditional (OR) expressions in the graph;

Author: No that you say it, I agree. I'm not sure how this section ended up in the User Guide. I'll just remove it, to a cylc-flow Issue...
